API design rule support

GeoSander opened the pull request:
# Overview
This PR adds an optional `api_rules` object to the `server` section of the pygeoapi configuration (YAML). For an example, please check the updated documentation [here](https://github.com/GeoSander/pygeoapi/blob/adr/docs/source/configuration.rst#server). Explanation about the available options can be found [here](https://github.com/GeoSander/pygeoapi/blob/adr/docs/source/configuration.rst#api-design-rules).

The suggested changes affect the API at the highest level (Flask, Starlette, Django). Because the current tests only test the lower level methods in `api.py` but not the routes leading towards them, I have added a [mock client](https://github.com/GeoSander/pygeoapi/blob/adr/tests/util.py#L79) for Flask  and Starlette so we can test API rule adherence or any routing-related things for that matter.

Note that I have also replaced the `@app.route()` decorators in favor of `Route()` in [starlette_app.py](https://github.com/geopython/pygeoapi/pull/1152/files#diff-22b78f4f30605039c367873f702cce35081fb85fe0590b04a9f432c0b43f0db0). This technique was deprecated anyway (it is not even documented anymore in the current Starlette docs) and it proved harder to build in the API rule logic like that.

# Related Issue / Discussion
This PR relates to issue #1134, which contains the rationale behind this change. 

# Additional Information
Since this work is part of a funded [tender](https://www.geonovum.nl/themas/invitation-to-tender) in which [Geonovum](https://www.geonovum.nl/about-us) emphasized that it was very important to them that the work would be incorporated into the core, and also because the suggested changes are optional and non-disruptive, we hope that the community is willing to approve this PR. 

# Contributions and Licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute feature "API design rule support" to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution.
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines.

GeoSander commented on pygeoapi/process/manager/tinydb_.py:
```diff
@@ -27,7 +27,12 @@
+try:
+    import fcntl
+except ModuleNotFoundError:
+    # When on Windows, fcntl does not exist and file locking is automatic
```
There really was no need to include this fix in this PR, but it made my life a lot easier (yes, I'm on Windows...), because the `ModuleNotFoundError` was raised on every API test.
For quicker testing, I do not wish to spin up a Linux Docker container all the time...

GeoSander commented on pygeoapi/util.py:
```diff
@@ -30,29 +30,29 @@
 import os
-from pathlib import Path
 import re
+from datetime import date, datetime, time
+from decimal import Decimal
+from enum import Enum
+from pathlib import Path
 from typing import Any, IO, Union
-from urllib.request import urlopen
+from typing import List
 from urllib.parse import urlparse
-
-from shapely.geometry import Polygon
+from urllib.request import urlopen
 
 import dateutil.parser
-from jinja2 import Environment, FileSystemLoader, select_autoescape
-from babel.support import Translations
 import yaml
+from babel.support import Translations
+from jinja2 import Environment, FileSystemLoader, select_autoescape
+from shapely.geometry import Polygon
```
I merely re-ordered the imports of built-ins and installed libs here. Looked a bit chaotic...

GeoSander commented on tests/pygeoapi-test-config-apirules.yml:
```diff
@@ -0,0 +1,345 @@
+# =================================================================
+#
+# Authors: Tom Kralidis <tomkralidis@gmail.com>
+#
+# Copyright (c) 2019 Tom Kralidis
```
Hmm... this should probably mention my name here... 🤷

GeoSander replied:
Hmm... I should probably mention my own name instead here... 🤷

tomkralidis commented on docs/source/configuration.rst:
```diff
@@ -60,6 +63,12 @@ The ``server`` section provides directives on binding and high level tuning.
+    api_rules:  # optional API design rules to which pygeoapi should adhere
+        api_version: 1.2.3  # omit to use pygeoapi's software version
+        strict_slashes: true  # trailing slashes will not be allowed and result in a 404
+        url_prefix: 'v{api_major}'  # adds a /v1 prefix to all URL paths
```
In theory, can this be any prefix (i.e. not starting with `v`)?

GeoSander replied:
@tomkralidis Correct. This is just an example. The docs elaborate on this.

GeoSander replied:
See [docs](https://github.com/geopython/pygeoapi/pull/1152/files#diff-e662b6389537d2a309e93ca2065f311d3f00156764a4835671300f6c4e95ec1dR274-R287) for what you can fill in here.

tomkralidis commented on pygeoapi/config.py:
```diff
@@ -40,6 +40,14 @@
 THISDIR = Path(__file__).parent.resolve()
 
 
+def load_json_schema() -> dict:
```
`s/load_json_schema/load_schema`?

GeoSander replied:
@tomkralidis Not sure what you mean by this?

I just put the code responsible for the actual schema loading out of `validate_config` and into its own function so I can reuse it (i.e. in the tests).

francbartoli replied:
@GeoSander I guess he's asking to change the function name to just `load_schema`

GeoSander replied:
Understood :)
Will rename it!

tomkralidis reviewed (changes requested):
See change requests.  Adding @francbartoli given the significant changes to starlette.

GeoSander commented:
Thanks for looking @tomkralidis. 
As for the "significant changes to Starlette" (@francbartoli), it's really not that significant except for the changes that were also done for Flask and Django. I mostly refactored the routing to follow a more fashionable approach, so I could better fit in the actual changes I needed to do.

francbartoli commented on tests/pygeoapi-test-config-apirules.yml:
```diff
@@ -0,0 +1,347 @@
+    bind:
+        host: 0.0.0.0
+        port: 5000
+    url: http://localhost:5000/
```
What does happen if this url has a root context, i.e. `http://localhost:5000/foo` and the `url_prefix` is set? Can you add a specific test for this?

GeoSander replied:
Should be fine, but I will (try to) add a test 👍

GeoSander replied:
@francbartoli I'm having trouble testing this. The mock client ignores the `url` setting in the YAML config completely.

In fact, if I set `url` to `http://localhost:5000/api` for example and spin up a Docker container, pygeoapi _does_ start successfully, but it actually runs at http:/localhost:5000 instead. All the links however do have the `/api` suffix, but if I click any of the links, none of them work. I have to remove the suffix to make them work.

The `url` property is only used in api.py and other code to create absolute URLs (which is not very dynamic actually...). Aside: I actually realize that I need to fix all those URL's too if a prefix applies...

**Q:** have **you** ever successfully set a root context using _only_ the `url` property?

In the case of Flask, I could make use of the built-in [`url_for`](https://flask.palletsprojects.com/en/2.2.x/api/?highlight=url_for#flask.url_for) function, which resolves the URL correctly, provided that I set the [`SERVER_NAME`](https://flask.palletsprojects.com/en/2.2.x/config/#SERVER_NAME) and [`APPLICATION_ROOT`](https://flask.palletsprojects.com/en/2.2.x/config/#APPLICATION_ROOT) correctly in the mock client, according to what's defined in pygeoapi's configuration.

francbartoli commented on tests/util.py:
```diff
@@ -67,5 +71,43 @@ def mock_request(params: dict = None, data=None, **headers) -> Request:
+
+@contextmanager
+def mock_client(config_file: str = 'pygeoapi-test-config.yml',
+                use_cookies: bool = True, **kwargs) -> FlaskClient:
```
Is it reasonable to have the same mock client and test for the other supported frameworks?

GeoSander replied:
I thought it was too much work for now (I'm on a budget...) so I focused on the default one. Tbh, I was actually disappointed that no one ever bothered to add mock clients for any framework. So I'd say this already is a win 😉

GeoSander replied:
Thinking.. perhaps I misunderstood you. You mean use the mock FlaskClient to also test Django and Starlette? For Starlette it might work (since that also uses Werkzeug, from which the FlaskClient inherits its mock client), for Django I'm not so sure.
We would also have to make any supported framework a dependency/requirement for the test and install that framework prior to running the tests.

francbartoli replied:
Well, actually I mean separate StarletteClient and DjangoClient

GeoSander replied:
Ah okay, so I did understand you correctly the first time 🙂 
I can see what I can do in the time available, focusing on Starlette first. 
Are these mock clients for all frameworks a "deal breaker" for you?

GeoSander replied:
Ok, I'll have a look later this week!

GeoSander commented on pygeoapi/models/config.py:
```diff
@@ -0,0 +1,95 @@
+
+    @staticmethod
+    def create(**rules_config) -> 'APIRules':
+        """ Returns a new APIDesignRules instance for the current API version
```
Yes, that docstring occurrence should be renamed.

GeoSander commented on docs/source/configuration.rst:
```diff
@@ -60,6 +63,12 @@ The ``server`` section provides directives on binding and high level tuning.
+        api_version: 1.2.3  # omit to use pygeoapi's software version
+        strict_slashes: true  # trailing slashes will not be allowed and result in a 404
+        url_prefix: 'v{api_major}'  # adds a /v1 prefix to all URL paths
+        version_header: X-API-Version  # add a response header of this name with the API version
```
Again, this is just an example. People may configure it to anyway they like (by default it won't be configured).
In the case of the Dutch API Design Rules (which I need to adhere to for my client Geonovum), they actually need it to be `API-Version`, but I preferred the `X-...` prefix here, which is what pygeoapi also does for the `X-Powered-By` header.

GeoSander commented:
Thanks for reviewing @francbartoli!

francbartoli commented on requirements-dev.txt:
```diff
@@ -9,7 +9,11 @@ pytest
 pytest-cov
 pytest-env
 coverage
-pyld
```
I guess this dependency is used in `tests/test_api.py`, the test is failing indeed https://github.com/geopython/pygeoapi/actions/runs/4396701156/jobs/7699332233

GeoSander replied:
Yes, I saw that it failed. Did not delete that dependency though. Or at least I wasn't aware that I did...
Anyway, will fix it of course 👍

GeoSander commented on pygeoapi/util.py:
```diff
@@ -135,6 +134,23 @@ class EnvVarLoader(yaml.SafeLoader):
+    return config_models.APIRules.create(**rules)
+
+
+def get_base_url(config: dict) -> str:
```
Added this function so we can get the complete base URL (with optional URL prefix) everywhere using the same code.

GeoSander commented on pygeoapi/util.py:
```diff
@@ -519,7 +534,7 @@ def url_join(*parts: list) -> str:
     """
 
-    return '/'.join([p.strip().strip('/') for p in parts])
+    return '/'.join([p.strip().strip('/') for p in parts]).rstrip('/')
```
While writing and testing `get_base_url()` above, I noticed that this function returned paths with trailing slashes if you are concatenating empty parts or parts that strip down to nothing.
The `rstrip('/')` should prevent this.

GeoSander commented:
Hi @tomkralidis / @francbartoli, I added the final changes (Starlette mock client, more tests) as per your request.

For Starlette, I ran into an interesting problem. Flask has a `strict_slashes=True` setting, which will **prevent** it from redirecting (308) to the URL without a trailing slash if the user added one and throws a 404 instead. Exactly what I needed. Starlette however has the `redirect_slashes` setting, but that only prevents the redirection (307) from happening and does not enforce strictness: adding a trailing slash will simply result in a 200 in that case, and in some cases [it still redirected anyway](https://github.com/encode/starlette/issues/869)...

So I have now added a [middleware class](https://github.com/geopython/pygeoapi/pull/1152/files#diff-22b78f4f30605039c367873f702cce35081fb85fe0590b04a9f432c0b43f0db0R472-R501) that tackles this problem. Note that the middleware is only installed when the `strict_slashes` API rule applies.

GeoSander commented:
Changed behavior of landing page requests according to OGC spec [here](https://docs.opengeospatial.org/is/17-069r4/17-069r4.html#_api_landing_page) if the `strict_slashes` API rule is in effect. For the landing page only, having a trailing slash is actually preferred (or mandatory even). 
Note that this actually contradicts the Dutch API Design Rules as discussed [here](https://gitlab.com/commonground/don/adr-validator/-/issues/5), which simply states that trailing slashes are never allowed.

GeoSander commented:
Update: rebased again to incorporate latests changes to `collection_items` route in starlette_app.py (added `OPTIONS` HTTP method).
